Fix #2937 - helm always appends /index.yaml at the end of URL

mcwienczek opened the pull request:
- closes #2937

mcwienczek committed:
Closes #2937
Added required dependency to run make test in developer's guide

mattfarina commented on docs/developers.md:
```diff
@@ -7,6 +7,7 @@ Helm and Tiller.
 
 - Go 1.6.0 or later
 - Glide 0.12.0 or later
+- Gometalinter
```
While this might not seem like much, what about updating the docs in a separate PR? One thing per change and this makes it two.

mcwienczek replied:
Removed.

mattfarina commented on pkg/downloader/chart_downloader.go:
```diff
@@ -125,6 +137,34 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+// https://github.com/kubernetes/helm/issues/2937
+// tries to fix URL because of BUG #2937
+// transforms URL from
+// https://url.to.repo/path?key=value/chart-0.1.0.tgz
```
How would we have gotten this broken url in the first place? What produces the broken url? Can that be fixed instead?

mcwienczek replied:
Removed the method and found the place where corrupt URL is constructed

mattfarina commented on pkg/downloader/chart_downloader.go:
```diff
@@ -125,6 +137,34 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+	appendedURL := chartURL[appendedURLIndex:]
+	restURL := chartURL[0:appendedURLIndex]
+
+	uu, _ := url.Parse(restURL)
```
Why is the error being skipped?

mcwienczek replied:
removed method

mattfarina commented on pkg/downloader/chart_downloader.go:
```diff
@@ -125,6 +137,34 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+// https://url.to.repo/path?key=value/chart-0.1.0.tgz
+// into
+// https://url.to.repo/path/chart-0-1-0.tgz?key=value
+func fixupURL(chartURL string) (string, error) {
```
This function makes a fair number of assumptions and takes a bit to understand what its doing. If there isn't another way to solve this problem this function needs to have good tests. I would hope there is an easier way to solve this.

mcwienczek replied:
Thank you. I have removed it and solved the problem at the source. I am not sure, though, if there is any other place that will be impacted.

mattfarina commented on pkg/downloader/chart_downloader.go:
```diff
@@ -125,6 +137,34 @@ func (c *ChartDownloader) DownloadTo(ref, version, dest string) (string, *proven
+	queryStringStartIndex := strings.Index(chartURL, "?")
+
+	if queryStringStartIndex == -1 {
+		return "", errors.New("Could not fixup URL")
```
Can this be a better error (if this function is kept)? How would someone know that a URL was broken and in need of fixing? And, what URL is causing that problem? I don't see how someone could take action based on this error to figure out how to correct the problem.

mcwienczek replied:
removed

mattfarina reviewed (changes requested):
Thanks for starting to work on this. It would be nice to have this fixed.

mcwienczek committed:
Fixed base URL appending when chart address is not absolute

mcwienczek committed:
Removed requirement from developers.md

mcwienczek committed:
Fixed unnecessary line breaks

mcwienczek committed:
Added tests for query string repo

mcwienczek committed:
Returning URL along with error

mcwienczek commented:
Thank you for review @mattfarina. I have removed `fixup` method and found the place where URL is constructed. I have also added one test. Please let me know if there is anything else I should do.

mattfarina reviewed (approved):
lgtm... though a real maintainer needs to come along and agree :)

technosophos commented:
As this is a bug fix, I am moving it to the next release (2.7) and merging it.
